File: 480. Sliding Window Median/Python/Sliding Window Median.py
class Solution(object):
    def medianSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[float]
        """
        min_heap = []
        max_heap = []
        result = []

        def get_median():
            if not max_heap and not min_heap:
                return 0
            if len(max_heap) == len(min_heap):
                return (min_heap[0] - max_heap[0]) / 2.0
            else:
                return float(min_heap[0])

        def add(num):
            if num < get_median():
                heapq.heappush(max_heap, -num)
            else:
                heapq.heappush(min_heap, num)
            if len(max_heap) > len(min_heap):
                heapq.heappush(min_heap, -heapq.heappop(max_heap))
            elif len(min_heap) > len(max_heap) + 1:
                heapq.heappush(max_heap, -heapq.heappop(min_heap))

        def remove(num):
            if num < get_median():
                max_heap.remove(-num)
                heapq.heapify(max_heap)
            else:
                min_heap.remove(num)
                heapq.heapify(min_heap)
            if len(max_heap) > len(min_heap):
                heapq.heappush(min_heap, -heapq.heappop(max_heap))
            elif len(min_heap) > len(max_heap) + 1:
                heapq.heappush(max_heap, -heapq.heappop(min_heap))

        for i in range(len(nums) + 1):
            if i >= k:
                result.append(get_median())
                remove(nums[i - k])
            if i < len(nums):
                add(nums[i])
        return result

File: 480. Sliding Window Median/C++/Sliding Window Median.cpp
class Solution {
public:
    vector<double> medianSlidingWindow(vector<int> &nums, int k) {
        class Utils {
            class custom_priority_queue : public std::priority_queue<long, std::vector < long>>

            {
                public:

                bool remove(const long &value) {
                    auto it = std::find(this->c.begin(), this->c.end(), value);
                    if (it != this->c.end()) {
                        this->c.erase(it);
                        std::make_heap(this->c.begin(), this->c.end(), this->comp);
                        return true;
                    } else {
                        return false;
                    }
                }
            };

        public:
            custom_priority_queue minHeap;
            custom_priority_queue maxHeap;

            double getMedian() {
                if (maxHeap.empty() && minHeap.empty()) {
                    return 0;
                }
                if (maxHeap.size() == minHeap.size()) {
                    return ((double) maxHeap.top() - minHeap.top()) / 2.0;
                }
                return -(double) minHeap.top();
            }

            void add(int num) {
                if (num < getMedian()) {
                    maxHeap.push((long) num);
                } else {
                    minHeap.push(-(long) num);
                }
                if (maxHeap.size() > minHeap.size()) {
                    minHeap.push(-maxHeap.top());
                    maxHeap.pop();
                } else if (minHeap.size() > maxHeap.size() + 1) {
                    maxHeap.push(-minHeap.top());
                    minHeap.pop();
                }
            }

            void remove(int num) {
                if (num < getMedian()) {
                    maxHeap.remove((long) num);
                } else {
                    minHeap.remove(-(long) num);
                }
                if (maxHeap.size() > minHeap.size()) {
                    minHeap.push(-maxHeap.top());
                    maxHeap.pop();
                } else if (minHeap.size() > maxHeap.size() + 1) {
                    maxHeap.push(-minHeap.top());
                    minHeap.pop();
                }
            }
        };
        vector<double> result;
        Utils utils;
        for (int i = 0; i <= nums.size(); i++) {
            if (i >= k) {
                double temp = utils.getMedian();
                temp = temp == 0 ? 0 : temp;
                result.push_back(temp);
            }
            if (i < nums.size()) {
                utils.add(nums[i]);
            }
        }
        return result;
    }
};